importers/file_meta.py

```diff
@@ -39,45 +40,111 @@ def _do_import(self, doc, filepath):
-    meta['Filename'] = meta['FileName']
+    meta['Filename'] = meta.get('FileName', '')
+    meta['Catalog'] = 'Exif'
+    meta['Height'] = meta.get('ImageHeight', '')
+    meta['Width'] = meta.get('ImageWidth', '')
     meta['ViewRotation'] = '0'
+    meta['CollectionId'] = self.collection_id
     meta['Rights'] = meta.get('UsageTerms', '')
     meta['ISOSpeedRating'] = meta.get('ISO', '')
-    '''
-    meta['ExposureBias'] = meta.get('ExposureBias', '')
-    meta['Aperture'] = meta.get('Aperture', '')
-    meta['MeteringMode'] = meta.get('MeteringMode', '')
-    meta['LightSource']  = meta.get('LightSource', '')
-    meta['Flash'] = meta.get('Flash', '')
-    meta['FocalLength'] = meta.get('FocalLength', '')
-    meta['SensingMethod'] = meta.get('SensingMethod', '')
-    meta['NoiseReduction'] = meta.get('NoiseReduction', '')
-    meta['Lens'] = meta.get('Lens', '')
-    '''
     meta['Latitude'] = meta.get('GPSLatitude')
     meta['Longitude'] = meta.get('GPSLongitude')
     meta['Altitude'] = meta.get('GPSAltitude')
     meta['SubjectReference'] = meta.get('SubjectCode')
     meta['Category'] = meta.get('SupplementalCategories')
     meta['Author'] = meta.get('Creator')
     meta['CaptureDate'] = meta.get('CreateDate')
+    meta['Model'] = "{} {}".format(meta.get('Make'), meta.get('Model'))
+
+    # Find Exif Headline field and use for "captions"
+    meta['Title'] = None if 'Title' not in meta else meta['Title'].strip()
     meta['Caption'] = []
     headlines = meta.get('Headline', '')
-    for line in headlines.split('\n'):
+    linelist = headlines.split("\n")
+    if len(linelist) < len(headlines.split("\r")):
+      linelist = headlines.split("\r")
+
+    for line in linelist:
       match = re.search('\{.+\[(.+)\]\}(.+)\{.+\[(.+)\]\}', line)
-      caption = None
       if match:
         lang_code = match.group(1)
         caption = match.group(2)
       else:
         lang_code = 'en'
         caption = line
+      if not meta['Title']:
+        meta['Title'] = caption
       meta['Caption'].append({
         'lang': lang_code,
         'text': caption
       })
 
+    # Add code for descriptions
+    descs = meta.get('Description')
+    if descs:
+      for line in descs.split('\n'):
+        match = re.search('\{.+\[(.+)\]\}(.+)\{.+\[(.+)\]\}', line)
+        if match:
+          lang_code = match.group(1)
+          desc = match.group(2)
+        else:
+          lang_code = 'en'
+          desc = line
+        foundit = False
+        for ind, cobj in enumerate(meta['Caption']):
+          if cobj['lang'] == lang_code:
+            meta['Caption'][ind]['description'] = desc
+            foundit = True
+        if not foundit:
+          meta['Caption'].append({
+            'lang': lang_code,
+            'text': 'Untitled',
+            'description': desc
+          })
+    meta['Caption'] = json.dumps(meta['Caption'])  # convert caption metadata to valid json
```

Comment:
Can you explain why the `json.dumps` is called on each of these items? Would it work if `meta` was a plain Python dict up until the end and then something like: `return json.dumps(meta)`?

Reply:
I tried many ways to get this to work and always lists within a field would wrongly get converted to JSON using single-quotes around everything, instead of escaping the double quotes. I thought at one point I sort of traced it back to the way the JSON was extracted by the `pyexifinfo` module, but that doesn't make sense. At any rate, when I tried doing what you suggest here, the PHP `json_decode` function, did not convert the list fields back into lists. Debugging it became a time sinkhole so I left it this way which works.

Reply:
Got it.